File: app/src/main/java/ru/evotor/framework/inventory/Measure.java
package ru.evotor.framework.inventory;

import android.os.Parcel;
import android.os.Parcelable;

/**
 * Created by a.kuznetsov on 30/04/2017.
 */

public class Measure implements Parcelable {
    private final String uuid;
    private final String name;
    private final int precision;

    public Measure(String uuid, String name, int precision) {
        this.uuid = uuid;
        this.name = name;
        this.precision = precision;
    }

    public String getUuid() {
        return uuid;
    }

    public String getName() {
        return name;
    }

    public int getPrecision() {
        return precision;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.uuid);
        dest.writeString(this.name);
        dest.writeInt(this.precision);
    }

    protected Measure(Parcel in) {
        this.uuid = in.readString();
        this.name = in.readString();
        this.precision = in.readInt();
    }

    public static final Parcelable.Creator<Measure> CREATOR = new Parcelable.Creator<Measure>() {
        @Override
        public Measure createFromParcel(Parcel source) {
            return new Measure(source);
        }

        @Override
        public Measure[] newArray(int size) {
            return new Measure[size];
        }
    };
}

File: src/main/java/ru/evotor/framework/receipt/Measure.kt
package ru.evotor.framework.receipt

import android.os.Parcel
import android.os.Parcelable
import ru.evotor.framework.ParcelableUtils
import ru.evotor.framework.kkt.FiscalRequisite
import ru.evotor.framework.kkt.FiscalTags

/**
 * Единица измерения.
 */
data class Measure(

        /**
         * Наименование
         */
        val name: String,
        /**
         * Точность
         */
        val precision: Int,
        /**
         * Код
         */
        @FiscalRequisite(tag = FiscalTags.MEASURE_CODE)
        val code: Int

) : Parcelable {

    override fun writeToParcel(dest: Parcel, flags: Int) {
        ParcelableUtils.writeExpand(dest, VERSION) { parcel ->
            parcel.writeString(name)
            parcel.writeInt(precision)
            parcel.writeInt(code)
        }
    }

    override fun describeContents(): Int = 0

    companion object {

        private const val VERSION = 1
        internal const val UNKNOWN_MEASURE_CODE = 255

        @JvmField
        val CREATOR = object : Parcelable.Creator<Measure> {
            override fun createFromParcel(parcel: Parcel): Measure = create(parcel)

            override fun newArray(size: Int): Array<Measure?> = arrayOfNulls(size)
        }

        private fun create(dest: Parcel): Measure {
            var measure: Measure? = null
            ParcelableUtils.readExpand(dest, VERSION) { parcel, version ->
                if (version >= 1) {
                    measure = Measure(
                            name = parcel.readString(),
                            precision = parcel.readInt(),
                            code = parcel.readInt()
                    )
                }
            }
            checkNotNull(measure)
            return measure as Measure
        }
    }
}